File: adapters/cdk/src/http_transport.rs
use std::time::Duration;

use async_trait::async_trait;
use reqwest::{Client, redirect::Policy};

use crate::funded::{DeliveryReceipt, TransportPort};

const MAX_RESPONSE_BYTES: usize = 65_536;

pub struct CdkHttpTransport {
    client: Client,
}

impl CdkHttpTransport {
    pub fn new(timeout: Duration) -> Result<Self, String> {
        let client = Client::builder()
            .redirect(Policy::none())
            .timeout(timeout)
            .build()
            .map_err(|_| "CDK HTTP transport initialization failed".to_owned())?;
        Ok(Self { client })
    }
}

#[async_trait]
impl TransportPort for CdkHttpTransport {
    async fn post(&self, target: &str, body: &[u8]) -> Result<DeliveryReceipt, String> {
        if body.len() > MAX_RESPONSE_BYTES {
            return Err("Cashu payment payload is too large".to_owned());
        }
        let mut response = self
            .client
            .post(target)
            .header("accept", "application/json")
            .header("content-type", "application/json")
            .body(body.to_vec())
            .send()
            .await
            .map_err(|error| {
                if error.is_timeout() {
                    "Cashu payment delivery timed out".to_owned()
                } else {
                    "Cashu payment delivery failed".to_owned()
                }
            })?;
        if response.status().is_redirection() {
            return Err("Cashu payment redirect is forbidden".to_owned());
        }
        if !response.status().is_success() {
            return Err(format!(
                "Cashu receiver returned HTTP {}",
                response.status()
            ));
        }
        if response
            .content_length()
            .is_some_and(|length| length > MAX_RESPONSE_BYTES as u64)
        {
            return Err("Cashu payment response is too large".to_owned());
        }
        let mut bytes = Vec::new();
        while let Some(chunk) = response
            .chunk()
            .await
            .map_err(|_| "Cashu payment response is invalid".to_owned())?
        {
            if bytes.len() + chunk.len() > MAX_RESPONSE_BYTES {
                return Err("Cashu payment response is too large".to_owned());
            }
            bytes.extend_from_slice(&chunk);
        }
        serde_json::from_slice(&bytes).map_err(|_| "Cashu payment response is invalid".to_owned())
    }
}

File: adapters/cashu-ts/src/http-transport.ts
import {
  validateAdapterResponse,
  type DeliveryReceiptView,
} from '@cashu-fault-lab/adapter-contract';
import type { CashuTsTransportPort, CashuTsTransportTarget } from './funded-operations.js';

const MAX_BODY_BYTES = 65_536;

export interface CashuTsHttpTransportOptions {
  readonly timeoutMs?: number;
  readonly fetch?: typeof fetch;
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new Error(`${name} must be a positive safe integer`);
  }
  return value;
}

function target(value: string): URL {
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new Error('Cashu payment target is invalid');
  }
  if (
    (url.protocol !== 'http:' && url.protocol !== 'https:') ||
    url.username.length > 0 ||
    url.password.length > 0 ||
    url.hash.length > 0
  ) {
    throw new Error('Cashu payment target is invalid');
  }
  return url;
}

async function boundedJson(response: Response): Promise<unknown> {
  const declared = response.headers.get('content-length');
  if (declared !== null && Number(declared) > MAX_BODY_BYTES) {
    throw new Error('Cashu payment response is too large');
  }
  if (response.body === null) throw new Error('Cashu payment response is empty');
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  while (true) {
    const next = await reader.read();
    if (next.done) break;
    length += next.value.byteLength;
    if (length > MAX_BODY_BYTES) {
      await reader.cancel();
      throw new Error('Cashu payment response is too large');
    }
    chunks.push(next.value);
  }
  const bytes = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    bytes.set(chunk, offset);
    offset += chunk.byteLength;
  }
  try {
    return JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(bytes)) as unknown;
  } catch {
    throw new Error('Cashu payment response is invalid');
  }
}

export class CashuTsHttpTransport implements CashuTsTransportPort {
  readonly #timeoutMs: number;
  readonly #fetch: typeof fetch;

  constructor(options: CashuTsHttpTransportOptions = {}) {
    this.#timeoutMs = positiveInteger(options.timeoutMs ?? 5_000, 'timeoutMs');
    this.#fetch = options.fetch ?? fetch;
  }

  async send(destination: CashuTsTransportTarget, body: Uint8Array): Promise<DeliveryReceiptView> {
    if (destination.type !== 'post') throw new Error('Cashu HTTP transport requires a POST target');
    return this.post(destination.target, body);
  }

  async post(destination: string, body: Uint8Array): Promise<DeliveryReceiptView> {
    if (!(body instanceof Uint8Array) || body.byteLength > MAX_BODY_BYTES) {
      throw new Error('Cashu payment payload is invalid');
    }
    const url = target(destination);
    const signal = AbortSignal.timeout(this.#timeoutMs);
    let response: Response;
    try {
      response = await this.#fetch(url, {
        method: 'POST',
        redirect: 'manual',
        signal,
        headers: { accept: 'application/json', 'content-type': 'application/json' },
        body,
      });
    } catch {
      if (signal.aborted) throw new Error('Cashu payment delivery timed out');
      throw new Error('Cashu payment delivery failed');
    }
    if (response.status >= 300 && response.status < 400) {
      await response.body?.cancel();
      throw new Error('Cashu payment redirect is forbidden');
    }
    if (!response.ok) {
      await response.body?.cancel();
      throw new Error(`Cashu receiver returned HTTP ${response.status}`);
    }
    const value = await boundedJson(response);
    const validation = validateAdapterResponse('delivery', value);
    if (!validation.ok) throw new Error('Cashu payment response violates the receipt contract');
    return value as DeliveryReceiptView;
  }
}
